test/multi-ns/connection_test.go

```diff
@@ -0,0 +1,159 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	clientDep = &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "simple-client",
+			Namespace: "default",
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: int32Ptr(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"networkservicemesh.io/app": "simple-client",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"networkservicemesh.io/app": "simple-client",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "alpine-img",
+							Image:           "alpine:latest",
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command:         []string{"tail", "-f", "/dev/null"},
+						},
+					},
+				},
+			},
+		},
+	}
+)
+
+func TestSingleClientMultiNS(t *testing.T) {
+	for testName, c := range map[string]struct {
+		networkServices []string
+		annotations     map[string]string
+	}{
+		"Connect to one NSE": {
+			[]string{"foo"},
+			map[string]string{"ns.networkservicemesh.io": "foo?app=vl3-nse-foo"},
+		},
+		"Connect to two NSEs": {
+			[]string{"foo", "bar"},
+			map[string]string{"ns.networkservicemesh.io": "foo?app=vl3-nse-foo,bar?app=vl3-nse-bar"},
+		},
+	} {
+		t.Logf("Running test case: %s", testName)
+		var deployments []appsv1.Deployment
+		//Install network service
+		for _, name := range c.networkServices {
+			fmt.Printf("Installing Network Service %s...\n", name)
+			os.Setenv("REMOTE_IP", *ipAddr)
+			os.Setenv("SERVICENAME", name)
+			cmd := exec.Command(*nsePath)
+			err := cmd.Run()
+			if err != nil {
+				t.Errorf("Failed to execute command: %v with error: %v", *nsePath, err.Error())
+				return
+			}
+		}
+		endpoints, err := listDeployments("wcm-system", clientset)
+		if err != nil {
+			t.Errorf("Failed to list Deployments in namespace '%v' with error: %v", "wcm-system", err.Error())
+			return
+		}
+		deployments = append(deployments, endpoints...)
+
+		//Create client pod with annotation
+		clientDep.ObjectMeta.Annotations = c.annotations
+		fmt.Println("Create client deployment")
+		deploymentsClient := clientset.AppsV1().Deployments(corev1.NamespaceDefault)
+		result, err := deploymentsClient.Create(context.TODO(), clientDep, metav1.CreateOptions{})
+		if err != nil {
+			t.Errorf("Failed to create Deployment '%v' with error: %v", clientDep.ObjectMeta.Name, err.Error())
+			return
+		}
+		deployments = append(deployments, *result)
+
+		var ready bool
+		for i := 0; i < 12; i++ {
+			fmt.Println("Waiting for client pod to start...")
+			time.Sleep(time.Duration(5) * time.Second)
+			ready = checkAvailability(result, clientset)
```

Comment:
I think a client pod is running doesn't necessarily mean it is connected to the `networkservice`, for example:
<img width="1615" alt="Screen Shot 2021-02-16 at 9 08 49 PM" src="https://user-images.githubusercontent.com/29384080/108216431-76ae3880-7100-11eb-9eb9-f087a0a96816.png">
In this case, the client pods are running, but they somehow don't have the `nsmd` and `nsmd-k8s` containers injected in the pods, and they are not connected to the `networkservice` although they are indeed in `Running` state.